src/resolver/descriptor.cc

```diff
@@ -0,0 +1,382 @@
+/* ========================================================================== *
+ *
+ * @file resolver/descriptor.cc
+ *
+ * @brief A set of user inputs used to set input preferences and query
+ *        parameters during resolution.
+ *
+ *
+ * -------------------------------------------------------------------------- */
+
+#include <regex>
+
+#include "versions.hh"
+#include "flox/core/exceptions.hh"
+#include "flox/resolver/descriptor.hh"
+
+
+/* -------------------------------------------------------------------------- */
+
+namespace flox::resolver {
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Sets either `version` or `semver` on
+ *        a `flox::resolver::ManifestDescriptor`.
+ *
+ * Distinguishes between semver ranges and exact version matchers.
+ * @param desc The descriptor to initialize.
+ * @param version The version description to parse.
+ */
+  static void
+initManifestDescriptorVersion(       ManifestDescriptor & desc
+                             , const std::string        & version
+                             )
+{
+  /* Strip leading/trailing whitespace. */
+  std::string trimmed = trim_copy( version );
+
+  /* Empty is recognized as a glob/_any_ range. */
+  if ( trimmed.empty() )
+    {
+      desc.semver = std::move( trimmed );
+      return;
+    }
+
+  /* Try a quick detection based on first character.
+   * We identify `=` as an explicit _exact version_ match. */
+  switch ( trimmed.at( 0 ) )
+    {
+      case '=':
+        desc.version = std::move( trimmed.substr( 1 ) );
+        break;
+
+      case '*':
+      case '~':
+      case '^':
+      case '>':
+      case '<':
+        desc.semver = std::move( trimmed );
+        break;
+
+      default:
+        /* If it's a valid semver or a date then it's not a range. */
+        if ( versions::isSemver( trimmed ) || versions::isDate( trimmed ) ||
+             ( ! versions::isSemverRange( trimmed ) )
+           )
+          {
+            desc.version = std::move( trimmed );
+          }
+        else /* Otherwise, assume a range. */
+          {
+            desc.semver = std::move( trimmed );
+          }
+        break;
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+/** @brief Get a `flox::resolver::AttrPathGlob` from a string if necessary. */
+  static AttrPathGlob
+maybeSplitAttrPathGlob( const ManifestDescriptorRaw::AbsPath & absPath )
+{
+  if ( std::holds_alternative<AttrPathGlob>( absPath ) )
+    {
+      return std::get<AttrPathGlob>( absPath );
+    }
+  AttrPathGlob   glob;
+  flox::AttrPath path = splitAttrPath( std::get<std::string>( absPath ) );
+  size_t         idx  = 0;
+  for ( const auto & part : path )
+    {
+      /* Treat `null' or `*' in the second element as a glob. */
+      if ( ( idx == 1 ) && ( ( part == "null" ) || ( part == "*" ) ) )
```

Comment:
Blocking if we're not throwing when we should

Reply:
IDK I have a bunch of error checks all over the place in here and a ton of tests so I think it's likely caught.
You can add more tests in a follow up though or recommend patches to improve the logic if you'd like.

Reply:
I have a struct definition like that in `github:flox/resolver` actually.
It just never made the migration because it felt like _another class_ that I only really needed in like 1 routine so I just in-lined it.

https://github.com/flox/resolver/blob/main/include/flox/types.hh#L115